### DISCUSSION TITLE
How do I update the view after not changing the viewmodel binding source?

### DISCUSSION THREAD

**Discussion by Justin113D** [AvaloniaUI/Avalonia#17131] [Q&A]
I have a `TextBox` with a `Mode=TwoWay` binding to a ViewModel and `UpdateSourceTrigger=LostFocus`. In the ViewModel, whitespace gets trimmed on setting, something like this:
```c#
        private string _text = string.Empty;

        public string Text
        {
            get => _text;
            set
            {
                _text = value.Trim();
                PropertyChanged?.Invoke(this, new(nameof(Text)));
            }
        }
```
In other words, upon losing focus, i want the `TextBox` to update and remove all whitespace at the start and end. The issue is that, when the only thing i do is add whitespace, the value behind the binding does not change, and so the text in the UI does not change either, meaning the whitespace remains.

I have no idea how to properly make it update and have it remove the whitespace after adding whitespace, but i did find a reliable workaround:
```c#
private string _workaroundText = string.Empty;

public string WorkaroundText
{
    get => _workaroundText;
    set
    {
        string previousValue = _workaroundText;
        _workaroundText = value.Trim();
        PropertyChanged?.Invoke(this, new(nameof(WorkaroundText)));

        if(previousValue == _workaroundText && value != _workaroundText)
        {
            string newValue = _workaroundText;
            _workaroundText += ' ';
            PropertyChanged?.Invoke(this, new(nameof(WorkaroundText)));
            _workaroundText = newValue;
            PropertyChanged?.Invoke(this, new(nameof(WorkaroundText)));
        }
    }
}
```

Essentially, i do make a change, tell the ui to update with it, change it back, and tell the UI to update again. This works, but is ugly. Is there a better solution?

Workaround in action:
![bindingWorkaround](https://github.com/user-attachments/assets/7045f1bd-77dd-44c3-93b8-9c22be860b69)

**Comment by PFox28** [comment]
I am actually facing the same problem: I have a setter of a property that corrects the user input. However, the UI only updates using this trick above.
Honestly, I find it ridiculous that the _normal_, straightforward apporach doesn't work here. In my opinion, setters of bound properties are made to control user input. I don't see the point why it's not reflected back to UI by default.

**Comment by LaurentInSeattle** [comment]
- Just derive a new control from TextBox and implement your custom lost focus behavior in there.

or even simpler: 

- write a couple of lines in code-behind.

No reason that the behavior of a control should be implemented in a ViewModel.

**Comment by PFox28** [comment]
I'm coming from WPF where the same code works as expected. 

Viewmodel:
```
    public class MainWindowViewModel : INotifyPropertyChanged
    {
        public event PropertyChangedEventHandler? PropertyChanged;

        public string Text
        {
            get;
            set
            {
                field = value.ToUpper();
                PropertyChanged?.Invoke(this, new(nameof(Text)));
            }
        }
        = "Beispiel";
    }
```

Binding:
`    <TextBox Text="{Binding Text, Mode=TwoWay, UpdateSourceTrigger=LostFocus}"/>
`

WPF:
<img width="150" height="100" alt="wpf" src="https://github.com/user-attachments/assets/db63d134-cfac-4c1f-8536-092389949f6f" />

Avalonia:
<img width="164" height="104" alt="avalonia" src="https://github.com/user-attachments/assets/b705e272-1122-4be4-849c-93388a90fa39" />

I don't even know exactly where this behavior in Avalonia differs from WPF. But I can see that it's not due to LostFocus. Even PropertyChanged produces confusing behavior in Avalonia, whereas in WPF it works as you'd expect: the text is always capitalized.
In Avalonia, the change is only triggered when you delete a character, but not if you add one. That's very confusing and most probably not intended.

WPF (PropertyChanged):
<img width="134" height="94" alt="wpf_2" src="https://github.com/user-attachments/assets/8ade3193-4681-484e-a68a-5546976ef894" />

Avalonia (PropertyChanged):
<img width="126" height="102" alt="avalonia_2" src="https://github.com/user-attachments/assets/ee4288cc-c3e4-40dc-98e7-eea80a7bcfb0" />
